### PR TITLE
Support CTE reuse rules

### PR DESCRIPTION
**Pull request by Seaven**
For #1318
* Support CTE produce column prune
* Support CTE consume predicate push down
* Support CTE consume limit push down
* Support CTE consume inline when one consume
* Support CTE consume inline by check costs

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java [hunk 1]
```diff
@@ -199,10 +207,12 @@
                 new PushDownPredicateIntersectRule(),
                 new PushDownPredicateTableFunctionRule(),
                 new PushDownPredicateRepeatRule(),
+
                 MergePredicateRule.HIVE_SCAN,
                 MergePredicateRule.SCHEMA_SCAN,
                 MergePredicateRule.MYSQL_SCAN,
-                new MergeTwoFiltersRule()
+                new MergeTwoFiltersRule(),
+                new PushDownPredicateCTEConsumeRule()
```
Comment (darionyaphet) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:215]: why some rules use static instances and another use new to constructe objects?

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:215]: Some static instances just for reuse code. In fact, many rule-list is singletons in `RuleSet`, so these ways are the same. 

Of course, we want to organize all the rules into singletons later.

Comment (darionyaphet) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:215]: It seems reasonable

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class CTEContext {
+    // All CTE produce
+    private Map<String, OptExpression> produces;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:18]: Why CTE id use string, not int?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java [hunk 3]
```diff
@@ -34,7 +34,8 @@ public MergeTwoProjectRule() {
             resultMap.put(entry.getKey(), entry.getValue().accept(rewriter, null));
         }
 
-        OptExpression optExpression = new OptExpression(new LogicalProjectOperator(resultMap));
+        OptExpression optExpression = new OptExpression(
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java:37]: Would better put this change to another PR, and merge it immediately.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushCTEProduceRule.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ *                    CTEProduce
+ *                         |
+ *  CTEProduce           Limit
+ *      |         =>       |
+ *     Node              Filter
+ *                         |
+ *                        Node
+ *
+ * */
+public class PushCTEProduceRule extends TransformationRule {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushCTEProduceRule.java:33]: PushDownLimitAndFilterToCTEProduceRule ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java [hunk 5]
```diff
@@ -93,11 +94,13 @@ public boolean check(OptExpression input, OptimizerContext context) {
         LogicalMetaScanOperator newMetaScan =
                 new LogicalMetaScanOperator(metaScan.getTable(), newScanColumnRefs, aggColumnIdToNames);
 
-        OptExpression newProject = new OptExpression(project);
-        newProject.getInputs().add(OptExpression.create(newMetaScan));
+        LogicalProjectOperator newProject = new LogicalProjectOperator(newScanColumnRefs.keySet().stream().collect(
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownAggToMetaScanRule.java:97]: Why need this change?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java [hunk 6]
```diff
@@ -388,29 +387,26 @@ public LogicalPlan visitTable(TableRelation node, ExpressionMapping context) {
 
     @Override
     public LogicalPlan visitCTE(CTERelation node, ExpressionMapping context) {
-        if (session.getSessionVariable().isCboCteReuse()) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:391]: Will always enable CteReuse?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneProjectRule.java [hunk 7]
```diff
@@ -56,7 +56,8 @@ public boolean check(OptExpression input, OptimizerContext context) {
                     logicalOperator.getOutputColumns(new ExpressionContext(input.inputAt(0))).getStream().
                             mapToObj(context.getColumnRefFactory()::getColumnRef).collect(Collectors.toList()));
             projectMap.put(smallestColumn, smallestColumn);
-            return Lists.newArrayList(OptExpression.create(new LogicalProjectOperator(projectMap), input.getInputs()));
+            return Lists.newArrayList(OptExpression
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PruneProjectRule.java:59]: Would better put this change to another PR, and merge it immediately.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:18]: Use int is OK, there use string just for simplicity, can be modify it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java (hunk 6 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:391]: Analyze will follow CTE process, will check the flag in rule rewrite phase, if disable CTEReuse, will inline all CTE query

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java (hunk 3 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/MergeTwoProjectRule.java:37]: It's OK

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/TPCHPlanTest.java [hunk 8]
```diff
@@ -69,6 +69,12 @@ public void testSetOperation() {
         runFileUnitTest("optimized-plan/set-operation");
     }
 
+    @Test
+    public void test() throws Exception {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/TPCHPlanTest.java:73]: Test case?

File: fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java [hunk 9]
```diff
@@ -741,6 +741,10 @@ public void setStatisticStorage(StatisticStorage statisticStorage) {
         this.statisticStorage = statisticStorage;
     }
 
+    public StatisticStorage getStatisticStorage() {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/Catalog.java:744]: It looks like unused

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 10]
```diff
@@ -163,6 +168,8 @@ void logicalRuleRewrite(Memo memo, TaskContext rootTaskContext) {
         ruleRewriteIterative(memo, rootTaskContext, new MergeTwoProjectRule());
         ruleRewriteOnlyOnce(memo, rootTaskContext, new PushDownAggToMetaScanRule());
 
+        cleanUpMemoGroup(memo);
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:171]: Push down predicate/project will produce a lot of useless group, it's will take ask memo slow, so clean it there

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 11]
```diff
@@ -182,22 +189,56 @@ void logicalRuleRewrite(Memo memo, TaskContext rootTaskContext) {
...
         ruleRewriteOnlyOnce(memo, rootTaskContext, LimitPruneTabletsRule.getInstance());
         ruleRewriteIterative(memo, rootTaskContext, RuleSetType.PRUNE_PROJECT);
+
+        cleanUpMemoGroup(memo);
+        if (cteContext.needOptimizeCTE()) {
+            CTEUtils.collectCteOperators(memo, context);
+        }
+
+        // compute CTE inline
+        while (cteContext.needOptimizeCTE() && cteContext.hasInlineCTE()) {
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:228]: Inline CTE on above accord by CTE consume nums, there will inline accord by cost, I will add comment

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 12]
```diff
@@ -182,22 +189,56 @@ void logicalRuleRewrite(Memo memo, TaskContext rootTaskContext) {
...
+            ruleRewriteIterative(memo, rootTaskContext, RuleSetType.INLINE_CTE);
+            CTEUtils.collectCteOperators(memo, context);
+        }
+
+        ruleRewriteIterative(memo, rootTaskContext, new MergeTwoProjectRule());
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:233]: The child plan of CTE consume maybe start with ProjectOperator, will produce two project after CTE consume inline

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+     */
+    public boolean needInline(String cteId) {
+        // 1. Disable CTE reuse
+        // 2. CTE consume only use once
+        if (!enableCTE || consumeNums.getOrDefault(cteId, 0) <= 1) {
+            return true;
+        }
+
+        // Wasn't collect costs, don't inline
+        if (produceCosts.isEmpty() && consumeInlineCosts.isEmpty()) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:164]: Or ？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        Preconditions.checkState(produceCosts.containsKey(cteId));
+        Preconditions.checkState(consumeInlineCosts.containsKey(cteId));
+
+        double p = produceCosts.get(cteId);
+        double c = consumeInlineCosts.get(cteId).stream().reduce(Double::sum).orElse(Double.MAX_VALUE);
+
+        // if was none statistic, default don't inline
+        if (p < 0 || c < 0) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:175]: costs could be LT 0？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:172]: Is it necessary ？ cleanUpMemoGroup is used many times in this function, it is time-consuming operation， what happened if do not add

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptimizerContext;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalCTEConsumeOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+
+public class InlineCTEConsumeRule extends TransformationRule {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java:15]: I think this rule is confusing， Whether it can be split into InlineOneCTEConsumeRule and PruneCTEConsumeRule？

File: fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java [hunk 16]
```diff
@@ -64,4 +64,6 @@
     public static String null_string = "\\N";
 
     public static boolean USE_MOCK_DICT_MANAGER = false;
+
+    public static double default_cte_inline_cost_rate = 1.5;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/common/FeConstants.java:68]: may be in StatisticsEstimateCoefficient is better

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java [hunk 17]
```diff
@@ -88,7 +88,7 @@ private ScalarOperator rewrite(ScalarOperator predicate, boolean onlyExtractColu
             }
         }
 
-        if (extractExpr != null && !conjuncts.contains(extractExpr)) {
+        if (!conjuncts.contains(extractExpr)) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java:91]: The necessary? extractExpr could always be non-null？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateCTEConsumeRule.java [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+/*
+ *          Filter          CTEConsume(Predicate)
+ *            |                 |
+ *        CTEConsume   =>     Filter
+ *            |                 |
+ *           Node              Node
+ *
+ * */
+public class PushDownPredicateCTEConsumeRule extends TransformationRule {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateCTEConsumeRule.java:24]: Whether need PushDownLimitCTEConsumeRule？

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownPredicateCTEConsumeRule.java:24]: Yes, you can find it in PushDownLimitDirectRule

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java (hunk 17 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarRangePredicateExtractor.java:91]: It's been checked before, don't need check again

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java (hunk 13 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:164]: Yeah, but I think and is strict

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java (hunk 14 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:175]: If none statistics?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:172]: Unnecessary, but when I test collect statistic, I found that useless groups take a lot of time for the whole task, as long as a few minutes, so I called cleanup many times. 
The useless groups produce by InlineCTE/PushdownPredicate/Limit may be several times normal groups.
Like some tpcds query, groups maybe 50+ before PushdownPredicate,  then will be 80+ after PushdownPredicate.

**Review by Youngwb — APPROVED** [review]

**Review by HangyuanLiu — APPROVED** [review]
